File: ServerSentEvents/resources/cors-cookie.php
<?php
  $id = isset($_SERVER['HTTP_LAST_EVENT_ID'])
  	? $_SERVER['HTTP_LAST_EVENT_ID']
	: '';

  $ident = isset($_GET['ident'])
  	? $_GET['ident']
	: 'test';

  $cookie = isset($_COOKIE[$ident])
  	? "COOKIE"
	: "NO_COOKIE";

  $origin = isset($_GET['origin'])
  	? $_GET['origin']
	: $_SERVER['HTTP_ORIGIN'];

  $credentials = isset($_GET['credentials'])
  	? $_GET['credentials']
	: 'true';

  if ($origin != 'none')
    header("Access-Control-Allow-Origin: {$origin}");

  if ($credentials != 'none')
    header("Access-Control-Allow-Credentials: {$credentials}");


  if ($id == '') {
    header("Content-Type: text/event-stream");
    header("Set-Cookie: $ident=COOKIE");

    echo "id: 1\n";
    echo "retry: 200\n";
    echo "data: first $cookie\n\n";
  }
  else if ($id == '1') {
    header("Content-Type: text/event-stream");
    header("Set-Cookie: $ident=COOKIE; expires=Fri, 27 Jul 2001 02:47:11 UTC");

    echo "id: 2\n";
    echo "data: second $cookie\n\n";
  }
  else {
    header("Content-Type: stop");
    echo "data: $id $cookie\n\n";
  }

File: eventsource/resources/cors-cookie.py
from datetime import datetime

from wptserve.utils import isomorphic_decode

def main(request, response):
    last_event_id = request.headers.get(b"Last-Event-Id", b"")
    ident = request.GET.first(b'ident', b"test")
    cookie = u"COOKIE" if ident in request.cookies else u"NO_COOKIE"
    origin = request.GET.first(b'origin', request.headers[b"origin"])
    credentials = request.GET.first(b'credentials', b'true')

    headers = []

    if origin != b'none':
        headers.append((b"Access-Control-Allow-Origin", origin));

    if credentials != b'none':
        headers.append((b"Access-Control-Allow-Credentials", credentials));

    if last_event_id == b'':
        headers.append((b"Content-Type", b"text/event-stream"))
        response.set_cookie(ident, b"COOKIE")
        data = u"id: 1\nretry: 200\ndata: first %s\n\n" % cookie
    elif last_event_id == b'1':
        headers.append((b"Content-Type", b"text/event-stream"))
        long_long_time_ago = datetime.now().replace(year=2001, month=7, day=27)
        response.set_cookie(ident, b"COOKIE", expires=long_long_time_ago)
        data = u"id: 2\ndata: second %s\n\n" % cookie
    else:
        headers.append((b"Content-Type", b"stop"))
        data = u"data: " + isomorphic_decode(last_event_id) + cookie + u"\n\n";

    return headers, data
